cpp/include/cudf/dictionary/dictionary_column_view.hpp

```diff
@@ -0,0 +1,235 @@
+ * d2 = remove_unused_dictionary(d1)
+ * d2 is now {["a","d"],[1,0,1,0]}
+ * ```
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> remove_unused_dictionary( dictionary_column_view const& dictionary_column,
+                                                  rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by applying only the specified dictionary elements
+ * to the existing dictionary_column.
+ *
+ * Any new elements found in the dictionary parameter are added to the output dictionary.
+ * Any existing dictionary elements not in the dictionary parameter are removed.
+ *
+ * The indices are update to reflect the position values of the new dictionary.
+ * Any indices pointing to removed elements are set to null.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = set_dictionary(d1,["a","b","c"])
+ * d2 is now {["a","b","c"],[null,0,2,0]}
+ * ```
+ *
+ * @throw cudf_logic_error if the dictionary type does not match the dictionary type in
+ * the dictionary_column.
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param dictionary New keys to use for the output column.
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> set_dictionary( dictionary_column_view const& dictionary_column,
+                                        column_view const& dictionary,
+                                        rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by merging the specified dictionary elements
+ * from two existing dictionary_columns.
+ *
+ * The indices are updated to appear as though the second dictionary_column is appended to the first.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = {["b","e"],[1,0,1,1,0]}
+ * d3 = merge(d1,d2)
+ * d3 is now {["a","b","c","d","e"],[3,0,2,0,4,1,4,4,1]}
+ * ```
+ *
+ * @throw cudf_logic_error if the dictionary types do not match for both dictionary columns.
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param dictionary New keys to remove from the dictionary_column
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> merge_dictionaries( dictionary_column_view const& dictionary_column1,
+                                            dictionary_column_view const& dictionary_column2,
+                                            rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by copying the dictionary from the provided
+ * dictionary_column and setting the indices from the provided indices parameter.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = gather(d1,[1,0,0,2,2])
```

Comment:
Why does this need to be called `gather`? All it is doing is making a new dictionary column using the dictionary from an existing column and a new set of indices. This role should be filled by a factory instead, shouldn't it?